File: scripts/0904_text_mining_and_nlp/04_assignment/assignment_01.py

# BACKGROUND: 
# 
# The given data contain short and crisp movie reviews by various critics.
# 
# QUESTIONS:
# 
#   1. Import Textdata. Do the essential cleaning of the data.
#   2. Find top 20 words sorted by frequency.
#   3. Create a wordcloud using the given data.
#   4. List the number of lines having sentiments ‘Negative', 'Neutral’ and ‘Positive’.
#   5. Plot bar graph showing top 15 words by frequency.





# %% 0 - import libraries
import pandas as pd
import matplotlib.pyplot as plt
import nltk

from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.sentiment.vader import SentimentIntensityAnalyzer

from wordcloud import WordCloud

from string import punctuation, digits










# 1. Import Textdata. Do the essential cleaning of the data.

# %% 1 - read data line by line...
with open("./data/0904_text_mining_and_nlp/04_assignment/Textdata.txt") as file:
    lines = [line.strip() for line in file]



# %% 1 - create data frame from lines...
data = pd.DataFrame(lines, columns= ["Review"])
data.head()
#                                               Review
# 0  films adapted from comic books have had plenty...
# 1  for starters , it was created by alan moore ( ...
# 2  to say moore and campbell thoroughly researche...
# 3  the book ( or " graphic novel , " if you will ...
# 4  in other words , don't dismiss this film becau...



# %% 1 - create function to clean the data...
stop_words    = set(stopwords.words('english'))
remove_punc   = str.maketrans('', '', punctuation)
remove_digits = str.maketrans('', '', digits)

def preprocess(line):
    line = line.lower()
    line = line.translate(remove_punc)
    line = line.translate(remove_digits)
    line = " ".join([word for word in word_tokenize(line) if word not in stop_words])
    return " ".join(line.strip().split())



# %% 1 - clean the data and assign it to column...
data["Cleaned_Review"] = data["Review"].apply(preprocess)
data.head()
#                                               Review  \
# 0  films adapted from comic books have had plenty...   
# 1  for starters , it was created by alan moore ( ...   
# 2  to say moore and campbell thoroughly researche...   
# 3  the book ( or " graphic novel , " if you will ...   
# 4  in other words , don't dismiss this film becau...   

#                                       Cleaned_Review  
# 0  films adapted comic books plenty success wheth...  
# 1  starters created alan moore eddie campbell bro...  
# 2  say moore campbell thoroughly researched subje...  
# 3  book graphic novel pages long includes nearly ...  
# 4                     words dont dismiss film source  



# %% 1 - aggregate all words into a single string - to be used later...
words = data["Cleaned_Review"].str.cat(sep = " ")










# 2. Find top 20 words sorted by frequency.

# %% 1 - create frequency distribution object and find 20 most common words...
freq_dist = nltk.FreqDist(word_tokenize(words))
freq_dist.most_common(20)
# [('film', 10),
#  ('like', 7),
#  ('dont', 7),
#  ('make', 7),
#  ('even', 6),
#  ('movie', 6),
#  ('comic', 5),
#  ('get', 5),
#  ('pretty', 5),
#  ('films', 4),
#  ('world', 4),
#  ('really', 4),
#  ('say', 4),
#  ('little', 4),
#  ('good', 4),
#  ('see', 4),
#  ('one', 4),
#  ('teen', 4),
#  ('book', 3),
#  ('moore', 3)]










# 3. Create a wordcloud using the given data.

# %% 1 - create wordcloud from words...
wordcloud = WordCloud(width = 800, height = 400, background_color = "white").generate(words)



# %% 1 - plot the wordcloud...
plt.figure(figsize = (10, 5))
plt.imshow(wordcloud, interpolation = "bilinear")
plt.axis("off")
plt.show()










# 4. List the number of lines having sentiments ‘Negative', 'Neutral’ and ‘Positive’.

# %% 1 - create instance of sentiment intensity analyzer...
sid = SentimentIntensityAnalyzer()



# %% 1 - get sentiment scores...
data["Sentiment_Scores"] = data["Cleaned_Review"].apply(lambda x: sid.polarity_scores(x))
data.iloc[:, 2].head()
# 0    {'neg': 0.19, 'neu': 0.632, 'pos': 0.178, 'com...
# 1    {'neg': 0.0, 'neu': 0.882, 'pos': 0.118, 'comp...
# 2    {'neg': 0.103, 'neu': 0.769, 'pos': 0.128, 'co...
# 3    {'neg': 0.0, 'neu': 0.796, 'pos': 0.204, 'comp...
# 4    {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...
# Name: Sentiment_Scores, dtype: object



# %% 1 - retrieve individual sentiment scores (we use compound scores)...
data["Positive_Score"] = data["Sentiment_Scores"].apply(lambda x: x["pos"])
data["Negative_Score"] = data["Sentiment_Scores"].apply(lambda x: x["neg"])
data["Neutral_Score"]  = data["Sentiment_Scores"].apply(lambda x: x["neu"])
data["Compound_Score"] = data["Sentiment_Scores"].apply(lambda x: x["compound"])
data.iloc[:, 3:].head()
#    Positive_Score  Negative_Score  Neutral_Score  Compound_Score
# 0           0.178           0.190          0.632         -0.1119
# 1           0.118           0.000          0.882          0.2500
# 2           0.128           0.103          0.769          0.1263
# 3           0.204           0.000          0.796          0.3182
# 4           0.000           0.000          1.000          0.0000



# %% 1 - lines wirth sentiment 'Negative' ...
data[(data["Compound_Score"] < 0)].shape[0]
# 21



# %% 1 - lines wirth sentiment 'Neutral' ...
data[(data["Compound_Score"] == 0)].shape[0]
# 19



# %% 1 - lines wirth sentiment 'Positive' ...
data[(data["Compound_Score"] > 0)].shape[0]
# 21










# 5. Plot bar graph showing top 15 words by frequency.

# %% 1 - create data frame with top 15 words, and their frequencies...
word_freq = pd.DataFrame(freq_dist.most_common(15), columns=["Word", "Freq"])
word_freq = word_freq.sort_values("Freq", ascending = True).reset_index(drop = True)



# %% 1 - plot bar chart of 15 words by their frequencies...
plt.figure(figsize = (16, 8))

plt.barh(word_freq.Word, word_freq.Freq, color = "cadetblue")

plt.title("Top 15 Words by Frequency")
plt.ylabel("Word")
plt.xlabel("Freq")

plt.show()

File: scripts/0904_text_mining_and_nlp/04_assignment/assignment_01.R

# BACKGROUND: 
# 
# The given data contain short and crisp movie reviews by various critics.
# 
# QUESTIONS:
# 
#   1. Import Textdata. Do the essential cleaning of the data.
#   2. Find words with minimum frequency 6.
#   3. List words with at least 0.35 correlation with ‘film’.
#   4. Create a wordcloud with words having minimum frequency 4. (Use any palette from RColorBrewer)
#   5. List the number of lines having sentiments ‘Sarcasm’, ‘Very Negative’ and ‘Very Positive’.
#   6. Plot graph showing words occurring more than 3 times (Use tidytext package).





library(tm)
library(wordcloud)
library(RSentiment)
library(tidyverse)










# 1. Import Textdata. Do the essential cleaning of the data.

lines <- readLines("./data/0904_text_mining_and_nlp/04_assignment/Textdata.txt", skipNul = TRUE)
data  <- data.frame(Review = lines)
head(data)
# 1 films adapted from comic books have had plenty of success , whether they're about superheroes ( batman , superman , spawn ) , or geared toward kids ( casper ) or the arthouse crowd ( ghost world ) , but there's never really been a comic book like from hell before . 
# 2                                                                                              for starters , it was created by alan moore ( and eddie campbell ) , who brought the medium to a whole new level in the mid '80s with a 12-part series called the watchmen . 
# 3                                                                                                                    to say moore and campbell thoroughly researched the subject of jack the ripper would be like saying michael jackson is starting to look a little odd . 
# 4                                                                                                                                the book ( or " graphic novel , " if you will ) is over 500 pages long and includes nearly 30 more that consist of nothing but footnotes . 
# 5                                                                                                                                                                                                          in other words , don't dismiss this film because of its source . 
# 6                                                                                                                              if you can get past the whole comic book thing , you might find another stumbling block in from hell's directors , albert and allen hughes .


data$Cleaned_Review <- tolower(data$Review)
data$Cleaned_Review <- removePunctuation(data$Cleaned_Review)
data$Cleaned_Review <- removeNumbers(data$Cleaned_Review)
data$Cleaned_Review <- removeWords(data$Cleaned_Review, stopwords("english"))
data$Cleaned_Review <- str_squish(data$Cleaned_Review)
head(data)
#                                                                        Review
# 1 films adapted from comic books have had plenty of success , whether they're about superheroes ( batman , superman , spawn ) , or geared toward kids ( casper ) or the arthouse crowd ( ghost world ) , but there's never really been a comic book like from hell before . 
# 2                                                                                              for starters , it was created by alan moore ( and eddie campbell ) , who brought the medium to a whole new level in the mid '80s with a 12-part series called the watchmen . 
# 3                                                                                                                    to say moore and campbell thoroughly researched the subject of jack the ripper would be like saying michael jackson is starting to look a little odd . 
# 4                                                                                                                                the book ( or " graphic novel , " if you will ) is over 500 pages long and includes nearly 30 more that consist of nothing but footnotes . 
# 5                                                                                                                                                                                                          in other words , don't dismiss this film because of its source . 
# 6                                                                                                                              if you can get past the whole comic book thing , you might find another stumbling block in from hell's directors , albert and allen hughes . 
#                                                                                                                                                              Cleaned_Review
# 1 films adapted comic books plenty success whether superheroes batman superman spawn geared toward kids casper arthouse crowd ghost world never really comic book like hell
# 2                                                               starters created alan moore eddie campbell brought medium whole new level mid s part series called watchmen
# 3                                                         say moore campbell thoroughly researched subject jack ripper like saying michael jackson starting look little odd
# 4                                                                                              book graphic novel will pages long includes nearly consist nothing footnotes
# 5                                                                                                                                                 words dismiss film source
# 6                                                                can get past whole comic book thing might find another stumbling block hells directors albert allen hughes










# 2. Find words with minimum frequency 6.

tdm <- TermDocumentMatrix(data$Cleaned_Review)
findFreqTerms(tdm, 6)
# [1] "like"  "dont"  "film"  "even"  "make"  "movie"










# 3. List words with at least 0.35 correlation with ‘film’.

findAssocs(tdm, "film", 0.35)
# $film
# biggest     now 
#    0.42    0.39


findAssocs(tdm, "movie", 0.35)
# $movie
#   mindfuck      sorta   critique generation    package   presents    touches   problems       lazy       mean    melissa      plain    running    showing    visions  different     giving    insight   offering     decent    decided 
#       0.56       0.39       0.39       0.39       0.39       0.39       0.39       0.39       0.39       0.39       0.39       0.39       0.39       0.39       0.39       0.39       0.39       0.39       0.39       0.39       0.39 
#       edge    mightve  somewhere      suits      video       teen 
#       0.39       0.39       0.39       0.39       0.39       0.36










# 4. Create a wordcloud with words having minimum frequency 4. (Use any palette from RColorBrewer)

words <- data %>%
  dplyr::select(Cleaned_Review) %>%
  tidytext::unnest_tokens(Word, Cleaned_Review) %>%
  dplyr::count(Word, sort = TRUE, name = "Freq") %>%
  dplyr::filter(Freq >= 4) %>%
  ungroup()

wordcloud(words$Word, words$Freq, random.order = FALSE, colors = brewer.pal(8, "Dark2"))


words <- data %>%
  dplyr::select(Cleaned_Review) %>%
  tidytext::unnest_tokens(Word, Cleaned_Review) %>%
  dplyr::count(Word, sort = TRUE, name = "Freq") %>%
  ungroup()

wordcloud(words$Word, words$Freq, random.order = FALSE, min.freq = 4, colors = brewer.pal(8, "Dark2"))










# 5. List the number of lines having sentiments ‘Sarcasm’, ‘Very Negative’ and ‘Very Positive’.

calculate_total_presence_sentiment(data$Review)
# [1,] "Sarcasm" "Negative" "Very Negative" "Neutral" "Positive" "Very Positive"
# [2,] "9"       "12"       "6"             "20"      "6"        "8"


calculate_total_presence_sentiment(data$Cleaned_Review)
# [1,] "Sarcasm" "Negative" "Very Negative" "Neutral" "Positive" "Very Positive"
# [2,] "0"       "13"       "11"            "18"      "9"        "10"










# 6. Plot graph showing words occurring more than 3 times (Use tidytext package).

words <- data %>%
  dplyr::select(Cleaned_Review) %>%
  tidytext::unnest_tokens(Word, Cleaned_Review) %>%
  dplyr::count(Word, sort = TRUE, name = "Freq") %>%
  dplyr::filter(Freq >= 4) %>%
  ungroup()


ggplot(data = words, aes(x = reorder(Word, Freq), y = Freq, fill = Word)) + 
  geom_bar(stat = "identity") + 
  labs(title = "Words Occurring More Than 3 Times", x = "Word", y = "Freq") +
  coord_flip()
